kernel/src/net/socket/ip/stream/listen.rs

```diff
@@ -1,103 +1,59 @@
 // SPDX-License-Identifier: MPL-2.0
 
 use aster_bigtcp::{
-    errors::tcp::ListenError,
-    iface::BindPortConfig,
-    socket::{RawTcpSetOption, TcpState, UnboundTcpSocket},
+    socket::{RawTcpOption, RawTcpSetOption},
     wire::IpEndpoint,
 };
-use ostd::sync::PreemptDisabled;
 
 use super::{connected::ConnectedStream, StreamObserver};
 use crate::{
     events::IoEvents,
-    net::iface::{BoundTcpSocket, Iface},
+    net::iface::{BoundPort, Iface, TcpListener},
     prelude::*,
-    process::signal::Pollee,
 };
 
 pub struct ListenStream {
-    backlog: usize,
-    /// A bound socket held to ensure the TCP port cannot be released
-    bound_socket: BoundTcpSocket,
-    /// Backlog sockets listening at the local endpoint
-    backlog_sockets: RwLock<Vec<BacklogSocket>, PreemptDisabled>,
+    tcp_listener: TcpListener,
 }
 
 impl ListenStream {
     pub fn new(
-        bound_socket: BoundTcpSocket,
+        bound_port: BoundPort,
         backlog: usize,
-        pollee: &Pollee,
-    ) -> core::result::Result<Self, (Error, BoundTcpSocket)> {
+        option: &RawTcpOption,
+        observer: StreamObserver,
+    ) -> Self {
         const SOMAXCONN: usize = 4096;
-        let somaxconn = SOMAXCONN.min(backlog);
+        let max_conn = SOMAXCONN.min(backlog);
 
-        let listen_stream = Self {
-            backlog: somaxconn,
-            bound_socket,
-            backlog_sockets: RwLock::new(Vec::new()),
+        let tcp_listener = match TcpListener::new_listen(bound_port, max_conn, option, observer) {
+            Ok(tcp_listener) => tcp_listener,
+            Err((_, err)) => {
+                unreachable!("`new_listen` fails with {:?}, which should not happen", err)
```

Comment:
`new_listen` can fail. On Linux, two sockets cannot listen on the same port, even if both sockets have `SO_REUSEADDR` set to true. This logic should be verified in `TcpListener::new_listen`, so this method can indeed fail.

Although we currently do not perform this check, it would be better to return an error rather than using unreachable in this case.

Reply:
That's a good point.

However, if I remove the `unreachable!`, I'll have to do a pattern match on `err` and report the error code as the result of the system call.

Then comes the problem. All patterns are currently unreachable because the reachable one you describe doesn't exist:
```rust
/// Error returned by [`Socket::listen`]
#[derive(Debug, PartialEq, Eq, Clone, Copy)]
#[cfg_attr(feature = "defmt", derive(defmt::Format))]
pub enum ListenError {
    InvalidState,
    Unaddressable,
}
```

So I would prefer to leave this as future work when we extend `ListenError` to include the error scenario you describe.

Reply:
It's fine to leave it as future work.